Collecting skill data (not complete)

oxisto opened the pull request:
Hi,

just to make you aware: I am currently working on collecting character's skill data from events. This is not complete yet, as the field that is probably related to Fame still puzzles me a little bit.

If you want, I can also rather open an issue an re-open a PR once the core is more stable.
In the meantime, I would love some feedback as i'm pretty inexperienced with Go.

The main issues that I still need to work on:
* I had to modify AlbionState to also include the Character Name, I dont know if this is the best idea. Looking at the pcaps, the character name is also included in lots of different events, maybe it would be better to look at those, rather than using the JoinOperation.
* As written above, the fame field is weird. From the first looks, it's neither max fame, nor fame for next level
* Currently, I have only arbitrary IDs of skills, would be nice to map them to actual skill names or at least the pseudo-names found in the game data, such as REFINE_T4 and so.

Let me know what you think :-)

Regner commented on client/operation_event_skill_data.go:
```diff
@@ -0,0 +1,55 @@
+package client
```
rename this file to `event_skill_data.go` please. :)

oxisto replied:
Ok. Should I go and introduce a new interface for events as well? currently i'm deriving from operation. I don't know what else this entails though.

oxisto replied:
Creating a new 'event' next to operation doesn't seem to be that easy. Specially this creates a problem in decode.go in decodeEvent() as it returns an Object with the interface operation. This gets then handed down to the router which also expects an operation interface. My suggestion would be that i rename the file for now, indicating that we will have multiple events in the future, but keep it an 'operation' internally for now.

Regner replied:
Agreed. :)

Regner commented on client/decode.go:
```diff
@@ -93,3 +93,21 @@ func decodeResponse(params map[string]interface{}) operation {
+	eventType := params["252"].(int16)
+
+	switch eventType {
+	case 114:
```
Please add the event to the wiki page for events as well: https://github.com/Regner/albiondata-client/wiki/Enums-Events

oxisto replied:
done

Regner commented on client/operation_join.go:
```diff
@@ -24,5 +24,7 @@ func (op operationJoinResponse) Process(state *albionState, uploader iuploader)
 	}
 
 	state.LocationId = loc
+	state.CharacterName = op.CharacterName
```
Can you add a debug line mentioning we are setting the character name, like the location one below.

Regner replied:
Can you check to see if this operation has the character ID as well as the character name? I would like that to get sent with private uploads as well as the name.

oxisto replied:
Sadly, it does not seem to have this information. I'm currently trying to look to the events to see if the ID can be found somewhere. I don't really get their ID system anyway... For example, I cannot seem to retrieve my id even using their API: https://gameinfo.albiononline.com/api/gameinfo/search?q=Oxisto yields nothing, although my character exists :) Searching for other characters work

Regner replied:
The character IDs are UUIDs, you can definitely see them when you look at the market JSON string.

Names are unique so it should be good enough for now, I was just really hoping that join response would contain the character ID. The request might though. If you feel like looking into it feel free, but I wont hold up this PR for it.

Regner commented on lib/nats.go:
```diff
@@ -7,4 +7,5 @@ const (
 	NatsMarketOrdersDeduped = "marketorders.deduped"
 	NatsMapDataIngest       = "mapdata.ingest"
 	NatsMapDataDeduped      = "mapdata.deduped"
+	NatsSkillDataIngest     = "skills"
```
Can you please adjust this block of consts to look more like this:

```go
// Public Topics
topic = "value"

// Private Topics
topic = "value"
```

Just to help keep it organized. :)

Regner commented:
First of all thank you very much for the PR! We really do appreciate it. I added a few comments, nothing crazy.

On the skill ID stuff it is probably similar to how they often do the item IDs. That being they load the contents of the XML files and then the ID is the index value of that item in the list of items from the XML.

Really really strange. We have not decided how we want to handle that. We don't really want to ship this static data with the client as it means we have to ship updates as soon as they release updates to Albion or we are potentially converting this information wrong and then sending the wrong information to the ingests.

Regner commented:
@pcdummy if you get a chance can you look the PR over.

oxisto commented:
Thanks for the comments, so far they all make sense. I'll look through them

oxisto commented:
Well, as it turns out, the information about character IDs is there, but in a very very weird way. I'll take the character of my wife as an example because mine seems to be bugged. As you already noted, you can see some UUID-style character ids in the market, they are 16-bytes represented as a hex string.

If you look at the "official" gameinfo API, you can see that character Ids suddently look like this: https://gameinfo.albiononline.com/api/gameinfo/players/tIy2OrzXQUi3jyllUPbZ-g.

While tIy2OrzXQUi3jyllUPbZ-g looks like base64, it is not 100 % base64, there is no minus in base64 and the alignment is missing. However, if you correct these minor things, you end up with tIy2OrzXQUi3jyllUPbZ+g==, which is proper base64 and can be represented as B48CB63ABCD74148B78F296550F6D9FA or B48CB63A-BCD7-4148-B78F-296550F6D9FA as the UUID-style you see in the market JSON. For starters, I don't know why they use two different representation of the same Id, it doesn't make any sense, but we need to decide which one we want to use.

Now if you look at what you can actually see in the JoinOperation (and probably also in other events) is: [58 -74 -116 -76 -41 -68 72 65 -73 -113 41 101 80 -10 -39 -6]. This an array of signed ints. However, these do not correspond with the real ID. For some reason you have to swap a few bytes (only in the first "half"):
0 -> 3, 3 -> 0
1 -> 2, 2 -> 1
4 -> 5, 5 -> 4
6 -> 7, 7 -> 6

Once you do that, you can use that information to either create the pseudo-base64 value or a uuid-style representation of that ID.

Phew.. probably wasted 2 hours of my life on such a weird ID system

Regner commented:
I just merged another PR into master. It shouldn't change much for you. The only real change you will need to make is how you pass the final message off to the ingest.

If you rebase to master and then check out one of the market operations you will see how that gets done now. :)

oxisto commented:
I rebased it and changed how i passed the message using sendMsgToPrivateUploaders. As far as I can see everything still works. Not sure if I have done the rebase correctly, I still had to merge it at the end. I never fully got the hang of git rebase tbh :D 

Any decision on the ID? Currently it is stored and sent in the pseudo-base64 format for my data.

oxisto commented:
Just for due diligence: So we found out, that Microsoft uses a special format to encode its UUID: https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding. That explains the byte-swapping. Re-organized the code a little bit and wrote LOTS of comments to understand it. Additionally the character id is now stored as UUID and as pseudo-base64, as discussed in discord.

Regner commented on lib/skills.go:
```diff
@@ -0,0 +1,17 @@
+type Skill struct {
+	ID                    int     `json:"Id"`
+	Level                 int     `json:"Level"`
+	PercentageOfNextLevel float64 `json:"PercentageOfNextLevel"`
```
What would you think of making this `PercentNextLevel` to be just a bit shorter.

Regner commented:
Looking really good, just two quick comments/questions and then I think we can merge.

oxisto commented:
Implemented your two comments. Should be "done" for now.

Regner commented:
Merged. Thank you very much for your work on this. It is greatly appreciated. :)
